### ISSUE TITLE
Refreshing state

### ISSUE THREAD

**Issue by makenosound** [atlassian-labs/react-resource-router#40]
I’ve _just_ started using this on a project and have a situation where I want to `refresh` the data after a change. At the moment it seems like there’s no easy way to differentiate these states:

* Loading for the first time
* Reloading after the initial fetch is successful

... and I wondered if there was scope to handle that in some way. Retaining the existing value of `data` would probably be enough I think (though I imagine that has flow-on effects that I am not seeing).

**Comment by rohan-deshpande** [comment]
Hey there, I might be misunderstanding but should there be a difference between your two points?

Currently `refresh` will always fetch the resource from remote meaning that it's a true refreshing of state. Could you possibly outline the differences you would expect to see between your two points in an application flow context?

Also, why would you want to retain the existing value of `data` if you also want to refresh that data?

**Comment by makenosound** [comment]
> but should there be a difference between your two points?

Yep, it is subtle, but there is a difference. The scenario I’m looking at is:

* A `listResource` loads 10 items sorted by date desc
* A user deletes one of those items (which is removed with an optimistic `update`)
* Refresh to fetch from the server with the latest data, which should gives us an additional item to replace the delete one.

With the current API there’s not an _easy_ way to differentiate between the `loading` values for the refresh vs the initial load. And ideally in the scenario I’m describing we’d have a different initial loading state to one when _most_ of the data is the same. Does that clarify?

A possible API could be to allow an argument to be passed to `refresh` that would retain any previous data until the new fetch completes.

**Comment by rohan-deshpande** [comment]
> A possible API could be to allow an argument to be passed to refresh that would retain any previous data until the new fetch completes.

So if I understand correctly you don't want the hooked/subscribed component to go back to the loading state while the refresh is occurring, rather, you'd like the current data to be displayed until the fetch has completed?

**Comment by makenosound** [comment]
> you'd like the current data to be displayed until the fetch has completed?

That is correct, yep. A "stale while revalidate" if you will.

**Comment by rohan-deshpande** [comment]
Right! Okay, I think we _did_ discuss this @albertogasparin, I think it shouldn't be too hard to add and I feel it might be a fairly common requirement from consumers.

**Comment by makenosound** [comment]
Awesome, thanks! It’s definitely something that can be done around the existing API, but it adds a fair bit of boilerplate and I can seem it being a behaviour I’d want to use regularly.
